### PR TITLE
[SPARK-32752][SQL]Fix parsing aliasing typed interval literal issue

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR fixes the issue for parsing aliasing typed interval values

e.g.  cases below should be all valid ones

```sql

         > select interval '1 day' abc;
Error in query:
Error parsing ' 1 day abc' to interval, unrecognized number 'abc'(line 1, pos 16)

== SQL ==
select interval '1 day' abc
----------------^^^
```

```
spark-sql> select interval '1 day' as abc;
Error in query:
Error parsing ' 1 day as' to interval, unrecognized number 'as'(line 1, pos 16)

== SQL ==
select interval '1 day' as abc
----------------^^^
```

Before this PR, these cases will be captured by grammar rule `multiUnitsInterval` but not `typedLiteral`, e.g. for interval '1 day' day, the value is `1 day` and the unit is day， which actually should be parsed as `((interval 1 day) as day)`

In this PR, I add a new rule "INTERVAL STRING AS? identifier? " to handle interval typed literals(maybe aliasing) and 1-stringvalue-unit interval together and before the rule `multiUnitsInterval`.

when the "AS" key is not null, the `STRING` must be the whole context of the interval whether it is valid or not, the identifier should be the alias name or else the "AS" itself as alias name.

when the "AS" key is null
    1 and when the `STRING` contains any alphabet, we can assume that the  `STRING` is the whole context of the interval whether it is valid or not. the identifier part will be alias name if present
    2 and when the `STRING` contains no alphabet, e.g. '1' day, '-1' day, '\t1' day, so the identifier here should be the unit part whether it is valid or not.

### Why are the changes needed?

bug fix

### Does this PR introduce _any_ user-facing change?

yes, the cases shown in part I. will be valid cases

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -818,6 +818,7 @@ primaryExpression
 
 constant
     : NULL                                                                                     #nullLiteral
+    | INTERVAL STRING AS? identifier?                                                          #maybeAliasedIntervalLiteral
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]: can you explain how the bug triggers?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -573,7 +573,7 @@ object IntervalUtils {
     var pointPrefixed: Boolean = false
 
     def trimToNextState(b: Byte, next: ParseState): Unit = {
-      if (b <= ' ') {
+      if (Character.isWhitespace(b)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:576]: This worths a new PR.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]: OK, I will update pr description after ci passes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:576]: OK, got it

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]: Before this PR, these cases will be captured by grammar rule `multiUnitsInterval` but not `typedLiteral`, e.g. for interval '1 day' day, the value is `1 day` and the unit is day， which actually should be parsed as `((interval 1 day) as day)`

In this PR, I add a new rule "INTERVAL STRING AS? identifier? " to handle interval typed literals and 1-value-unit interval together and before the rule `multiUnitsInterval`.

when the "AS" key is not null, the `STRING` must be the whole context of the interval whether it is valid or not, the identifier should be the alias name or else the "AS" itself as alias name.

when the "AS" key is null
    1 and when the `STRING` contains any alphabet, we can assume that the  `STRING` is the whole context of the interval whether it is valid or not. the identifier part will be alias name if present
    2 and when the `STRING` contains no alphabet, e.g. '1' day, '-1' day, '\t1' day, so the identifier here should be the unit part whether it is valid or not.

Also updated in PR desc.

Comment (cloud-fan) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]: can you check `interval '1 day' day` in other systems? To see if they also treat '1 day' as the number of days and fail.

Comment (yaooqinn) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]:
```
postgres=# select interval '1 day' day;
 interval
----------
 1 day
(1 row)

postgres=# select interval '1' day day;
ERROR:  syntax error at or near "day"
LINE 1: select interval '1' day day;
                                ^
postgres=# select interval '1' day;
 interval
----------
 1 day
(1 row)
```

Comment (yaooqinn) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]:
```sql
presto> select interval '1 day' day;
Query 20200903_015817_00008_gec8v failed: Invalid INTERVAL DAY value: 1 day

presto> select interval '1' day day;
      day
----------------
 1 00:00:00.000
(1 row)

Query 20200903_015821_00009_gec8v, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto> select interval '1' day;
     _col0
----------------
 1 00:00:00.000
(1 row)

Query 20200903_015825_00010_gec8v, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto>
```

Comment (yaooqinn) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:821]: presto does not support typed literal for intervals

**Comment by yaooqinn** [comment]
retest this please

**Pull request closed without merging**